### PR TITLE
Add futures page

### PR DESCRIPTION
**Pull request by djmitche**
This adds a page about futures, showing how `async` and `await` interact with them.

### CODE REVIEW

File: src/async/futures.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+pub enum Poll<T> {
+    Ready(T),
+    Pending,
+}
+```
+
+An async function returns an `impl Future`, and an async block evaluates to an
+`impl Future`. It's also possible (but uncommon) to implement `Future` for your
+own types. For example, the `JoinHandle` returns from `tokio::spawn` implements
```
Comment (rbehjati) [inline src/async/futures.md:22]:
```suggestion
own types. For example, the `JoinHandle` returned from `tokio::spawn` implements
```

I think!

File: src/async/futures.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+The `.await` keyword, applied to a Future, causes the current async function or
+block to pause until that Future is ready, and then evaluates to its output.
+
+<details>
+
+* These types are conceptually quite simple, and implemented as such in
+  `std::task`.
+
+* We will get to `Pin` and `Context` shortly, so don't get into any depth on
```
Comment (rbehjati) [inline src/async/futures.md:44]: Cool. So you actually want to talk about `Pin` too.

**Review by rbehjati — APPROVED** [review]
Thanks :)

File: src/async/futures.md (hunk 2 above)
Comment (djmitche) [reply src/async/futures.md:44]: Based on discussion in the issue, I'm not sure it's a good idea, actually.  The ecosystem is _almost_ to the point where understanding Pin is only required for people working on low-level plumbing, and the rest of us can happily type `.await` and `async` and assume things just work.

Maybe I'll refocus this to a "how futures work, simplified" page and not revisit it in more detail.

Comment (rbehjati) [reply src/async/futures.md:44]: I like that. But maybe @mgeisler has other ideas :)

Comment (mgeisler) [reply src/async/futures.md]: I'm happy to wait a bit and see where the three of you take this — you know more about async than me!

**Pull request merged**